File: interfaces/vscode/src/taskProtocol.ts
/**
 * SaCode 任务协议（Rust kernel task_protocol.rs 的 TypeScript 精确映射）。
 *
 * 协议原则：旧格式可读、新格式只写；`yolo` 仅作为输入别名，客户端统一输出 `auto`；
 * 未知协议边界使用 `unknown` + 运行时校验，解析失败安全降级为 null，不崩溃、不误报成功。
 */

export const TASK_PROTOCOL_VERSION = 1;

export type EntrySource =
    | 'cli'
    | 'tui'
    | 'repl'
    | 'vscode'
    | 'automation'
    | 'sdk'
    | 'daemon';

export type ExecutionMode = 'auto' | 'plan' | 'build';

/** 可接受的输入模式：`yolo` 仅为旧值别名，规范化为 `auto` */
export type ExecutionModeInput = ExecutionMode | 'yolo';

export type TaskState =
    | 'pending'
    | 'ready'
    | 'retrying'
    | 'running'
    | 'waiting_for_user'
    | 'waiting_for_approval'
    | 'cancelling'
    | 'completed'
    | 'failed'
    | 'cancelled';

export type TaskPhase =
    | 'queued'
    | 'executing'
    | 'waiting_for_user'
    | 'waiting_for_approval'
    | 'cancelling'
    | 'finished';

export type TerminalOutcome = 'success' | 'failure' | 'cancelled';

export type FailureCategory =
    | 'provider'
    | 'permission'
    | 'approval'
    | 'tool'
    | 'validation'
    | 'recovery'
    | 'compatibility'
    | 'system';

export type SuggestedAction =
    | 'retry'
    | 'reconfigure_provider'
    | 'request_approval'
    | 'check_permissions'
    | 'update_client'
    | 'review_recovery_source'
    | 'report_issue'
    | 'none';

export type ValidationStatus = 'passed' | 'failed' | 'not_run' | 'unknown';

export interface FailureDetail {
    category: FailureCategory;
    code: string;
    retryable: boolean;
    suggested_action: SuggestedAction;
    safe_message: string;
}

export interface WorkspaceBoundary {
    root: string;
}

export interface ExplicitContextRef {
    path: string;
    start_line?: number;
    end_line?: number;
    character_count?: number;
}

export interface TaskCreateRequest {
    schema_version: number;
    prompt: string;
    mode: ExecutionMode;
    source: EntrySource;
    workspace: WorkspaceBoundary;
    explicit_contexts: ExplicitContextRef[];
}

export interface ResultSummary {
    text?: string;
    changed_objects: string[];
}

export interface RouteSummary {
    provider: string;
    model: string;
    switched: boolean;
    reason?: string;
}

export interface ValidationSummary {
    status: ValidationStatus;
    checks: string[];
}

export interface TaskTimestamps {
    created_at?: string;
    started_at?: string;
    updated_at?: string;
    finished_at?: string;
}

export interface TaskSnapshot {
    schema_version: number;
    task_id: string;
    mode: ExecutionMode;
    source: EntrySource;
    state: TaskState;
    phase: TaskPhase;
    terminal_outcome?: TerminalOutcome;
    failure?: FailureDetail;
    result?: ResultSummary;
    validation?: ValidationSummary;
    route?: RouteSummary;
    timestamps: TaskTimestamps;
}

const ENTRY_SOURCES: readonly EntrySource[] = [
    'cli', 'tui', 'repl', 'vscode', 'automation', 'sdk', 'daemon',
];

const TASK_STATES: readonly TaskState[] = [
    'pending', 'ready', 'retrying', 'running', 'waiting_for_user',
    'waiting_for_approval', 'cancelling', 'completed', 'failed', 'cancelled',
];

const FAILURE_CATEGORIES: readonly FailureCategory[] = [
    'provider', 'permission', 'approval', 'tool', 'validation',
    'recovery', 'compatibility', 'system',
];

const SUGGESTED_ACTIONS: readonly SuggestedAction[] = [
    'retry', 'reconfigure_provider', 'request_approval', 'check_permissions',
    'update_client', 'review_recovery_source', 'report_issue', 'none',
];

const VALIDATION_STATUSES: readonly ValidationStatus[] = [
    'passed', 'failed', 'not_run', 'unknown',
];

export function normalizeExecutionMode(value: unknown): ExecutionMode {
    if (value === 'yolo') return 'auto';
    return value === 'plan' || value === 'auto' || value === 'build'
        ? value
        : 'auto';
}

/** 与 Rust `TaskPhase::from(TaskState)` 一致的状态→阶段映射 */
export function phaseForState(state: TaskState): TaskPhase {
    switch (state) {
        case 'pending':
        case 'ready':
        case 'retrying':
            return 'queued';
        case 'running':
            return 'executing';
        case 'waiting_for_user':
            return 'waiting_for_user';
        case 'waiting_for_approval':
            return 'waiting_for_approval';
        case 'cancelling':
            return 'cancelling';
        case 'completed':
        case 'failed':
        case 'cancelled':
            return 'finished';
    }
}

export function isTerminalState(state: TaskState): boolean {
    return state === 'completed' || state === 'failed' || state === 'cancelled';
}

function terminalOutcomeFor(state: TaskState): TerminalOutcome | undefined {
    switch (state) {
        case 'completed': return 'success';
        case 'failed': return 'failure';
        case 'cancelled': return 'cancelled';
        default: return undefined;
    }
}

function isString(value: unknown): value is string {
    return typeof value === 'string';
}

function isOptionalString(value: unknown): value is string | undefined {
    return value === undefined || value === null || typeof value === 'string';
}

function isOptionalBoolean(value: unknown): value is boolean | undefined {
    return value === undefined || value === null || typeof value === 'boolean';
}

function oneOf<T extends string>(value: unknown, allowed: readonly T[]): T | null {
    return allowed.includes(value as T) ? (value as T) : null;
}

function isProtocolVersion(value: unknown): value is number {
    return typeof value === 'number' && Number.isInteger(value);
}

/** 校验协议版本：0 视为旧协议，>1 视为未知新协议，均不受支持 */
export function isProtocolVersionSupported(version: unknown): boolean {
    return isProtocolVersion(version) && version === TASK_PROTOCOL_VERSION;
}

function parseTimestamps(value: unknown): TaskTimestamps | null {
    if (value !== undefined && value !== null && typeof value !== 'object') return null;
    const raw = (value ?? {}) as Record<string, unknown>;
    const fields = ['created_at', 'started_at', 'updated_at', 'finished_at'] as const;
    const timestamps: TaskTimestamps = {};
    for (const field of fields) {
        if (!isOptionalString(raw[field])) return null;
        const text = raw[field];
        if (typeof text === 'string') timestamps[field] = text;
    }
    return timestamps;
}

function parseFailureDetail(value: unknown): FailureDetail | null {
    if (value === null || value === undefined) return null;
    if (typeof value !== 'object') return null;
    const raw = value as Record<string, unknown>;
    const category = oneOf(raw.category, FAILURE_CATEGORIES);
    if (!category || !isString(raw.code) || !isOptionalBoolean(raw.retryable)) return null;
    const suggestedAction = oneOf(raw.suggested_action, SUGGESTED_ACTIONS);
    if (!suggestedAction || !isString(raw.safe_message)) return null;
    return {
        category,
        code: raw.code,
        retryable: raw.retryable === true,
        suggested_action: suggestedAction,
        safe_message: raw.safe_message,
    };
}

function parseResultSummary(value: unknown): ResultSummary | null {
    if (value === null || value === undefined) return null;
    if (typeof value !== 'object') return null;
    const raw = value as Record<string, unknown>;
    if (!isOptionalString(raw.text)) return null;
    const changed = raw.changed_objects;
    if (changed !== undefined && !Array.isArray(changed)) return null;
    if (Array.isArray(changed) && !changed.every(isString)) return null;
    return {
        ...(typeof raw.text === 'string' ? { text: raw.text } : {}),
        changed_objects: Array.isArray(changed) ? changed : [],
    };
}

function parseRouteSummary(value: unknown): RouteSummary | null {
    if (value === null || value === undefined) return null;
    if (typeof value !== 'object') return null;
    const raw = value as Record<string, unknown>;
    if (!isString(raw.provider) || !isString(raw.model) || !isOptionalBoolean(raw.switched)) return null;
    if (!isOptionalString(raw.reason)) return null;
    return {
        provider: raw.provider,
        model: raw.model,
        switched: raw.switched === true,
        ...(typeof raw.reason === 'string' ? { reason: raw.reason } : {}),
    };
}

function parseValidationSummary(value: unknown): ValidationSummary | null {
    if (value === null || value === undefined) return null;
    if (typeof value !== 'object') return null;
    const raw = value as Record<string, unknown>;
    const status = oneOf(raw.status, VALIDATION_STATUSES);
    if (!status) return null;
    const checks = raw.checks;
    if (checks !== undefined && !Array.isArray(checks)) return null;
    if (Array.isArray(checks) && !checks.every(isString)) return null;
    return { status, checks: Array.isArray(checks) ? checks : [] };
}

/**
 * 运行时校验并解析 TaskSnapshot。
 *
 * 对应 Rust `TaskSnapshot::validate()` 的语义；任何不匹配（未知枚举、
 * phase 与 state 不一致、终态矛盾、failure 矛盾、schema 不支持）返回 null，
 * 调用方按降级提示处理，不得误报成功。
 */
export function parseTaskSnapshot(value: unknown): TaskSnapshot | null {
    if (value === null || value === undefined || typeof value !== 'object') return null;
    const raw = value as Record<string, unknown>;
    if (!isProtocolVersionSupported(raw.schema_version)) return null;
    if (!isString(raw.task_id) || raw.task_id.trim().length === 0) return null;
    const mode = oneOf(raw.mode, ['auto', 'plan', 'build'] as const);
    const source = oneOf(raw.source, ENTRY_SOURCES);
    const state = oneOf(raw.state, TASK_STATES);
    const phase = oneOf(raw.phase, [
        'queued', 'executing', 'waiting_for_user', 'waiting_for_approval', 'cancelling', 'finished',
    ] as const);
    if (!mode || !source || !state || !phase) return null;
    if (phase !== phaseForState(state)) return null;
    if (raw.terminal_outcome !== undefined && raw.terminal_outcome !== null) {
        if (raw.terminal_outcome !== terminalOutcomeFor(state)) return null;
    } else if (isTerminalState(state)) {
        return null;
    }
    const failure = parseFailureDetail(raw.failure);
    if ((state === 'failed') !== (failure !== null)) return null;
    const result = parseResultSummary(raw.result);
    const validation = parseValidationSummary(raw.validation);
    const route = parseRouteSummary(raw.route);
    const timestamps = parseTimestamps(raw.timestamps);
    if (!timestamps) return null;

    const snapshot: TaskSnapshot = {
        schema_version: raw.schema_version as number,
        task_id: raw.task_id,
        mode,
        source,
        state,
        phase,
        timestamps,
    };
    const outcome = terminalOutcomeFor(state);
    if (outcome) snapshot.terminal_outcome = outcome;
    if (failure) snapshot.failure = failure;
    if (result) snapshot.result = result;
    if (validation) snapshot.validation = validation;
    if (route) snapshot.route = route;
    return snapshot;
}

export const TASK_STATE_LABELS: Readonly<Record<TaskState, string>> = {
    pending: '排队中',
    ready: '待执行',
    retrying: '重试中',
    running: '执行中',
    waiting_for_user: '等待输入',
    waiting_for_approval: '等待审批',
    cancelling: '取消中',
    completed: '已完成',
    failed: '失败',
    cancelled: '已取消',
};

export const FAILURE_CATEGORY_LABELS: Readonly<Record<FailureCategory, string>> = {
    provider: '模型服务',
    permission: '权限',
    approval: '审批',
    tool: '工具',
    validation: '验证',
    recovery: '恢复',
    compatibility: '兼容性',
    system: '系统',
};

export const SUGGESTED_ACTION_LABELS: Readonly<Record<SuggestedAction, string>> = {
    retry: '重试任务',
    reconfigure_provider: '重新配置模型服务',
    request_approval: '请求审批',
    check_permissions: '检查权限',
    update_client: '升级扩展',
    review_recovery_source: '检查恢复源',
    report_issue: '报告问题',
    none: '无需操作',
};

File: kernel/src/schema/task_protocol.rs
use serde::{Deserialize, Serialize};
use std::fmt;

use super::{ExecutionMode, TaskState};

pub const TASK_PROTOCOL_VERSION: u32 = 1;

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum EntrySource {
    Cli,
    Tui,
    Repl,
    Vscode,
    Automation,
    Sdk,
    Daemon,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum TaskPhase {
    Queued,
    Executing,
    WaitingForUser,
    WaitingForApproval,
    Cancelling,
    Finished,
}

impl From<TaskState> for TaskPhase {
    fn from(state: TaskState) -> Self {
        match state {
            TaskState::Pending | TaskState::Ready | TaskState::Retrying => Self::Queued,
            TaskState::Running => Self::Executing,
            TaskState::WaitingForUser => Self::WaitingForUser,
            TaskState::WaitingForApproval => Self::WaitingForApproval,
            TaskState::Cancelling => Self::Cancelling,
            TaskState::Completed | TaskState::Failed | TaskState::Cancelled => Self::Finished,
        }
    }
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum TerminalOutcome {
    Success,
    Failure,
    Cancelled,
}

impl TerminalOutcome {
    pub fn task_state(self) -> TaskState {
        match self {
            Self::Success => TaskState::Completed,
            Self::Failure => TaskState::Failed,
            Self::Cancelled => TaskState::Cancelled,
        }
    }
}

impl TryFrom<TaskState> for TerminalOutcome {
    type Error = TaskProtocolError;

    fn try_from(state: TaskState) -> Result<Self, Self::Error> {
        match state {
            TaskState::Completed => Ok(Self::Success),
            TaskState::Failed => Ok(Self::Failure),
            TaskState::Cancelled => Ok(Self::Cancelled),
            _ => Err(TaskProtocolError::NonTerminalState(state)),
        }
    }
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum FailureCategory {
    Provider,
    Permission,
    Approval,
    Tool,
    Validation,
    Recovery,
    Compatibility,
    System,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum SuggestedAction {
    Retry,
    ReconfigureProvider,
    RequestApproval,
    CheckPermissions,
    UpdateClient,
    ReviewRecoverySource,
    ReportIssue,
    None,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct FailureDetail {
    pub category: FailureCategory,
    pub code: String,
    pub retryable: bool,
    pub suggested_action: SuggestedAction,
    pub safe_message: String,
}

impl FailureDetail {
    pub fn system(code: impl Into<String>, safe_message: impl Into<String>) -> Self {
        Self {
            category: FailureCategory::System,
            code: code.into(),
            retryable: false,
            suggested_action: SuggestedAction::ReportIssue,
            safe_message: safe_message.into(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct WorkspaceBoundary {
    pub root: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct ExplicitContextRef {
    pub path: String,
    #[serde(default)]
    pub start_line: Option<u32>,
    #[serde(default)]
    pub end_line: Option<u32>,
    #[serde(default)]
    pub character_count: Option<usize>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct TaskCreateRequest {
    pub schema_version: u32,
    pub prompt: String,
    pub mode: ExecutionMode,
    pub source: EntrySource,
    pub workspace: WorkspaceBoundary,
    #[serde(default)]
    pub explicit_contexts: Vec<ExplicitContextRef>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct ResultSummary {
    #[serde(default)]
    pub text: Option<String>,
    #[serde(default)]
    pub changed_objects: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct RouteSummary {
    pub provider: String,
    pub model: String,
    pub switched: bool,
    #[serde(default)]
    pub reason: Option<String>,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ValidationStatus {
    Passed,
    Failed,
    NotRun,
    Unknown,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct ValidationSummary {
    pub status: ValidationStatus,
    #[serde(default)]
    pub checks: Vec<String>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq, Eq)]
pub struct TaskTimestamps {
    #[serde(default)]
    pub created_at: Option<String>,
    #[serde(default)]
    pub started_at: Option<String>,
    #[serde(default)]
    pub updated_at: Option<String>,
    #[serde(default)]
    pub finished_at: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct TaskSnapshot {
    pub schema_version: u32,
    pub task_id: String,
    pub mode: ExecutionMode,
    pub source: EntrySource,
    pub state: TaskState,
    pub phase: TaskPhase,
    #[serde(default)]
    pub terminal_outcome: Option<TerminalOutcome>,
    #[serde(default)]
    pub failure: Option<FailureDetail>,
    #[serde(default)]
    pub result: Option<ResultSummary>,
    #[serde(default)]
    pub validation: Option<ValidationSummary>,
    #[serde(default)]
    pub route: Option<RouteSummary>,
    pub timestamps: TaskTimestamps,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct TaskFinalization {
    pub state: TaskState,
    pub applied: bool,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum TaskProtocolError {
    LegacySchema,
    UnsupportedSchema(u32),
    EmptyPrompt,
    EmptyTaskId,
    EmptyWorkspace,
    NonTerminalState(TaskState),
    StatePhaseMismatch,
    TerminalMismatch {
        state: TaskState,
        outcome: Option<TerminalOutcome>,
    },
    FailureMismatch,
}

impl fmt::Display for TaskProtocolError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::LegacySchema => write!(formatter, "legacy task protocol is unsupported"),
            Self::UnsupportedSchema(version) => {
                write!(formatter, "unsupported task protocol version: {version}")
            }
            Self::EmptyPrompt => write!(formatter, "task prompt cannot be empty"),
            Self::EmptyTaskId => write!(formatter, "task id cannot be empty"),
            Self::EmptyWorkspace => write!(formatter, "workspace root cannot be empty"),
            Self::NonTerminalState(state) => {
                write!(formatter, "task state is not terminal: {state}")
            }
            Self::StatePhaseMismatch => write!(formatter, "task phase does not match task state"),
            Self::TerminalMismatch { state, outcome } => {
                write!(
                    formatter,
                    "terminal outcome {outcome:?} does not match state {state}"
                )
            }
            Self::FailureMismatch => {
                write!(formatter, "failure detail does not match task outcome")
            }
        }
    }
}

impl std::error::Error for TaskProtocolError {}

impl TaskCreateRequest {
    pub fn validate(&self) -> Result<(), TaskProtocolError> {
        validate_protocol_version(self.schema_version)?;
        if self.prompt.trim().is_empty() {
            return Err(TaskProtocolError::EmptyPrompt);
        }
        if self.workspace.root.trim().is_empty() {
            return Err(TaskProtocolError::EmptyWorkspace);
        }
        Ok(())
    }
}

impl TaskSnapshot {
    pub fn validate(&self) -> Result<(), TaskProtocolError> {
        validate_protocol_version(self.schema_version)?;
        if self.task_id.trim().is_empty() {
            return Err(TaskProtocolError::EmptyTaskId);
        }
        if self.phase != TaskPhase::from(self.state) {
            return Err(TaskProtocolError::StatePhaseMismatch);
        }
        let expected_outcome = if self.state.is_terminal() {
            Some(TerminalOutcome::try_from(self.state)?)
        } else {
            None
        };
        if self.terminal_outcome != expected_outcome {
            return Err(TaskProtocolError::TerminalMismatch {
                state: self.state,
                outcome: self.terminal_outcome,
            });
        }
        if (self.state == TaskState::Failed) != self.failure.is_some() {
            return Err(TaskProtocolError::FailureMismatch);
        }
        Ok(())
    }
}

fn validate_protocol_version(version: u32) -> Result<(), TaskProtocolError> {
    match version {
        0 => Err(TaskProtocolError::LegacySchema),
        TASK_PROTOCOL_VERSION => Ok(()),
        other => Err(TaskProtocolError::UnsupportedSchema(other)),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn request() -> TaskCreateRequest {
        TaskCreateRequest {
            schema_version: TASK_PROTOCOL_VERSION,
            prompt: "inspect the repository".to_string(),
            mode: ExecutionMode::Build,
            source: EntrySource::Cli,
            workspace: WorkspaceBoundary {
                root: ".".to_string(),
            },
            explicit_contexts: Vec::new(),
        }
    }

    fn snapshot(state: TaskState) -> TaskSnapshot {
        TaskSnapshot {
            schema_version: TASK_PROTOCOL_VERSION,
            task_id: "task-1".to_string(),
            mode: ExecutionMode::Build,
            source: EntrySource::Cli,
            state,
            phase: TaskPhase::from(state),
            terminal_outcome: TerminalOutcome::try_from(state).ok(),
            failure: (state == TaskState::Failed)
                .then(|| FailureDetail::system("system/unknown", "Task failed")),
            result: None,
            validation: None,
            route: None,
            timestamps: TaskTimestamps::default(),
        }
    }

    #[test]
    fn task_create_request_accepts_yolo_only_as_input_alias() {
        let mut value = serde_json::to_value(request()).unwrap();
        value["mode"] = serde_json::json!("yolo");
        let parsed: TaskCreateRequest = serde_json::from_value(value).unwrap();
        assert_eq!(parsed.mode, ExecutionMode::Yolo);
        assert_eq!(serde_json::to_value(parsed).unwrap()["mode"], "auto");
    }

    #[test]
    fn task_create_request_rejects_empty_prompt() {
        let mut value = request();
        value.prompt = "  ".to_string();
        assert_eq!(value.validate(), Err(TaskProtocolError::EmptyPrompt));
    }

    #[test]
    fn task_snapshot_roundtrip_preserves_terminal_outcome() {
        for state in [
            TaskState::Completed,
            TaskState::Failed,
            TaskState::Cancelled,
        ] {
            let expected = snapshot(state);
            let serialized = serde_json::to_string(&expected).unwrap();
            let actual: TaskSnapshot = serde_json::from_str(&serialized).unwrap();
            assert_eq!(actual, expected);
            assert!(actual.validate().is_ok());
        }
    }

    #[test]
    fn task_snapshot_rejects_multiple_terminal_meanings() {
        let mut value = snapshot(TaskState::Completed);
        value.failure = Some(FailureDetail::system("system/unknown", "Task failed"));
        assert_eq!(value.validate(), Err(TaskProtocolError::FailureMismatch));
    }

    #[test]
    fn task_snapshot_rejects_unknown_enums() {
        let mut value = serde_json::to_value(snapshot(TaskState::Running)).unwrap();
        value["source"] = serde_json::json!("unknown");
        assert!(serde_json::from_value::<TaskSnapshot>(value).is_err());
    }
}
